StdlibUnittest: port the test harness to Windows

compnerd opened the pull request:
<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd committed:
stdlib: port StdlibUnittest to Windows

troughton commented on stdlib/private/StdlibUnittest/StdlibUnittest.swift:
```diff
@@ -749,12 +756,23 @@ func _stdlib_getline() -> String? {
+  var buffer: UnsafeMutablePointer<CChar>?
+  var length: Int = 0
+  if _dupenv_s(&buffer, &length, "SWIFT_INTERPRETER") != 0 {
+    if buffer != nil {
```
Very minor drive-by code-style: this might be cleaner as `if let buffer = buffer`. The two `if` conditions could also be folded together:

```
if _dupenv_s(&buffer, &length, "SWIFT_INTERPRETER") != 0, let buffer = buffer {
```

compnerd committed:
StdlibUnittest: add a workraound for DOS lineendings

Windows uses DOS line endings which appear through the PIPE endpoints.
We currently split on line feeds only, leaving the carriage return in
place.  Without this, the messages on the communication pipe would not
be interpreted correctly.

compnerd committed:
StdlibUnittest: prevent abort dialog on Windows

Adjust the abort behaviour on Windows.  The aborts would cause a large
number of dialogs to appear.  This reduces the load a bit and is crucial
for getting tests to run without manual intervention.

compnerd committed:
StdlibUnittest: add a VEH for TRAP on Windows

Add a vectored exception handler for illegal instructions on Windows.
This allows us to emulate `signal(SIGTRAP, ...)`.  This allows better
coverage of tests.

compnerd commented:
@swift-ci please test

compnerd commented:
CC: @xwu @moiseev

compnerd commented:
@swift-ci please test

gottesmm commented:
@swift-ci please test

compnerd commented:
CC: @moiseev

compnerd commented:
CC: @lorentey

moiseev reviewed (approved):
LGTM
